indexing-service/src/main/java/io/druid/indexing/common/task/TaskMonitor.java

```diff
@@ -0,0 +1,502 @@
+                    monitorEntry.setLastStatus(taskStatus);
+                    iterator.remove();
+                    break;
+                  case FAILED:
+                    incrementNumFailedTasks();
+
+                    log.warn("task[%s] failed!", taskId);
+                    if (monitorEntry.numTries() < maxRetry) {
+                      log.info(
+                          "We still have chances[%d/%d] to complete for spec[%s].",
+                          monitorEntry.numTries(),
+                          maxRetry,
+                          monitorEntry.spec.getId()
+                      );
+                      retry(specId, monitorEntry, taskStatus);
+                    } else {
+                      log.error(
+                          "spec[%s] failed after [%d] tries",
+                          monitorEntry.spec.getId(),
+                          monitorEntry.numTries()
+                      );
+                      // Remote the current entry after updating taskHistories to make sure that task history
+                      // exists either runningTasks or taskHistories.
+                      monitorEntry.setLastStatus(taskStatus);
+                      iterator.remove();
+                    }
+                    break;
+                  case RUNNING:
+                    monitorEntry.updateStatus(taskStatus);
+                    break;
+                  default:
+                    throw new ISE("Unknown taskStatus[%s] for task[%s[", taskStatus.getState(), taskId);
+                }
+              }
+            }
+          }
+          catch (Throwable t) {
+            log.error(t, "Error while monitoring");
+            throw t;
+          }
+        },
+        taskStatusCheckingPeriod,
+        taskStatusCheckingPeriod,
+        TimeUnit.MILLISECONDS
+    );
+  }
+
+  public void stop()
+  {
+    running = false;
+    taskStatusChecker.shutdownNow();
+    log.info("Stopped taskMonitor");
+  }
+
+  public ListenableFuture<SubTaskCompleteEvent<T>> submit(SubTaskSpec<T> spec)
+  {
+    if (!running) {
+      return Futures.immediateFailedFuture(new ISE("TaskMonitore is not running"));
+    }
+    final T task = spec.newSubTask(0);
+    log.info("Submitting a new task[%s] for spec[%s]", task.getId(), spec.getId());
+    indexingServiceClient.runTask(task);
+    incrementNumRunningTasks();
+
+    final SettableFuture<SubTaskCompleteEvent<T>> taskFuture = SettableFuture.create();
+    runningTasks.put(
+        spec.getId(),
+        new MonitorEntry(spec, task, indexingServiceClient.getTaskStatus(task.getId()).getStatus(), taskFuture)
+    );
+
+    return taskFuture;
+  }
+
+  /**
+   * Submit a retry task for a failed spec. This method should be called inside of the
+   * {@link java.util.concurrent.Callable} executed by {@link #taskStatusChecker}.
+   */
+  private void retry(String subTaskSpecId, MonitorEntry monitorEntry, TaskStatusPlus lastFailedTaskStatus)
+  {
+    if (running) {
```

Comment:
there's no state lock here. What if `stop` is called during this process?

Reply:
Good point. I'll add a lock.

Reply:
Added a lock.